enterprise/com/src/main/java/org/neo4j/com/storecopy/ToFileStoreWriter.java

```diff
@@ -23,46 +23,77 @@
+    private final List<FileMoveAction> fileMoveActions; // Todo need thread safe?
 
-    public ToFileStoreWriter( File graphDbStoreDir, StoreCopyClient.Monitor monitor )
+    public ToFileStoreWriter( File graphDbStoreDir, StoreCopyClient.Monitor monitor, PageCache pageCache,
+            List<FileMoveAction> fileMoveActions )
     {
         this.basePath = graphDbStoreDir;
         this.monitor = monitor;
+        this.pageCache = pageCache;
+        this.fileMoveActions = fileMoveActions;
     }
 
     @Override
-    public long write( String path, ReadableByteChannel data, ByteBuffer temporaryBuffer,
-            boolean hasData ) throws IOException
+    public long write( String path, ReadableByteChannel data, ByteBuffer temporaryBuffer, boolean hasData,
+            int recordSize ) throws IOException
     {
         try
         {
             temporaryBuffer.clear();
             File file = new File( basePath, path );
-
             file.getParentFile().mkdirs();
+
+            String filename = file.getName();
+            final Optional<StoreType> storeType = isStoreType( filename );
+            final Optional<PagedFile> existingMapping = pageCache.getExistingMapping( file );
+
             monitor.startReceivingStoreFile( file );
-            try ( RandomAccessFile randomAccessFile = new RandomAccessFile( file, "rw" ) )
+            try
             {
-                long totalWritten = 0;
-                if ( hasData )
+                if ( existingMapping.isPresent() )
                 {
-                    FileChannel channel = randomAccessFile.getChannel();
-                    while ( data.read( temporaryBuffer ) >= 0 )
+                    try ( PagedFile pagedFile = existingMapping.get() )
                     {
-                        temporaryBuffer.flip();
-                        totalWritten += temporaryBuffer.limit();
-                        channel.write( temporaryBuffer );
-                        temporaryBuffer.clear();
+                        final long written = writeDataThroughPageCache( pagedFile, data, temporaryBuffer, hasData );
+                        addPageCacheMoveAction( file );
+                        return written;
                     }
                 }
-                return totalWritten;
+                if ( storeType.isPresent() && storeType.get().isRecordStore() )
+                {
+                    int filePageSize = filePageSize( recordSize );
+                    try ( PagedFile pagedFile = pageCache.map( file, filePageSize, CREATE, WRITE ) )
+                    {
+                        final long written = writeDataThroughPageCache( pagedFile, data, temporaryBuffer, hasData );
+                        addPageCacheMoveAction( file );
+                        return written;
+                    }
+                }
+                else
+                {
+
+                    return writeDataThroughFileSystem( file, data, temporaryBuffer, hasData );
```

Comment:
Why don't this add a move action? Then we'd only need one loop in `StoreCopyClient`.